File: Go/Balanced Binary Tree.go
// 4ms, 98.77%; 5.7MB, 100.00%
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
	return helper(root) >= 0
}

func helper(node *TreeNode) int {
	if node == nil {
		return 0
	}
	lHeight, rHeight := helper(node.Left), helper(node.Right)
	if lHeight < 0 || rHeight < 0 || abs(lHeight-rHeight) > 1 {
		return -1
	}
	return 1 + max(lHeight, rHeight)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: Java/Balanced Binary Tree.java
// 0ms, 100.00%; 39.5MB, 73.15%
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public static boolean isBalanced(TreeNode root) {
        return helper(root) > -1;
    }

    private static int helper(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int lheight = helper(node.left), rheight = helper(node.right);
        if (lheight < 0 || rheight < 0 || Math.abs(lheight - rheight) > 1) {
            return -1;
        }
        return 1 + Math.max(lheight, rheight);
    }
}
